From e9db509e4c4d7df9725d79f692a51d629dfd5fba Mon Sep 17 00:00:00 2001
From: Fancy Fang <chen.fang@nxp.com>
Date: Fri, 20 Oct 2017 18:44:23 +0800
Subject: [PATCH 2717/5242] MLK-16706-1 video: fbdev: mipi_dsi_northwest:
 replace 'PICOS2KHZ' by 'PICOS2KHZ2'

commit  90121e2cb7a43cfe664635cd933c4ea3c62eafa2 from
https://source.codeaurora.org/external/imx/linux-imx.git

The 'PICOS2KHZ' macro is used to get pixel clock frequency
from 'pixclock' value to derive the required mipi phy bit
clock frequency. But the result precision get from this
macro is not good enough in some cases. The patch defines
an new improved macro 'PICOS2KHZ2' to replace 'PICOS2KHZ'.

Signed-off-by: Fancy Fang <chen.fang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/video/fbdev/mxc/mipi_dsi_northwest.c |    6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
index 278e55e..d4dbff7 100644
--- a/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
+++ b/drivers/video/fbdev/mxc/mipi_dsi_northwest.c
@@ -52,6 +52,9 @@
 #define NS2PS_RATIO			(1000)
 #define	MIPI_LCD_SLEEP_MODE_DELAY	(120)
 #define MIPI_FIFO_TIMEOUT		msecs_to_jiffies(250)
+#define PICOS_PER_SEC			(1000000000UL)
+#define PICOS2KHZ2(a, bpp)		\
+	DIV_ROUND_CLOSEST(PICOS_PER_SEC * (bpp), (a))
 
 static struct mipi_dsi_match_lcd mipi_dsi_lcd_db[] = {
 #ifdef CONFIG_FB_MXC_TRULY_WVGA_SYNC_PANEL
@@ -301,7 +304,8 @@ static int mipi_dsi_dphy_init(struct mipi_dsi_info *mipi_dsi)
 #endif
 
 	bpp = fmt_to_bpp(lcd_config->dpi_fmt);
-	req_bit_clk = PICOS2KHZ(mode->pixclock) * bpp * 1000U;
+	req_bit_clk = PICOS2KHZ2(mode->pixclock, bpp) * 1000U;
+
 	switch (lcd_config->data_lane_num) {
 	case 1:
 		break;
-- 
1.7.9.5

